crates/script-sequence/src/reader.rs

```diff
@@ -0,0 +1,142 @@
+use crate::{ScriptSequence, TransactionWithMetadata};
+use alloy_rpc_types::AnyTransactionReceipt;
+use eyre::{bail, Context, Result};
+use foundry_common::fs;
+use revm_inspectors::tracing::types::CallKind;
+use std::path::{Path, PathBuf};
+
+/// This type reads broadcast files in the
+/// `project_root/broadcast/{contract_name}.s.sol/{chain_id}/` directory.
+///
+/// It consists methods that filter and search for transactions in the broadcast files that match a
+/// `transactionType` if provided.
+///
+/// Note:
+///
+/// It only returns transactions for which there exists a corresponding receipt in the broadcast.
+#[derive(Debug, Clone)]
+pub struct BroadcastReader {
+    contract_name: String,
+    chain_id: u64,
+    tx_type: Option<CallKind>,
+    broadcast_path: PathBuf,
+}
+
+impl BroadcastReader {
+    /// Create a new `BroadcastReader` instance.
+    pub fn new(contract_name: String, chain_id: u64, root: &Path) -> Result<Self> {
+        let broadcast_path = root
+            .join("broadcast")
+            .join(format!("{contract_name}.s.sol"))
+            .join(chain_id.to_string());
```

Comment:
looks like `contract_name` here is actually treated as script contract name? I'd expect this to go over all existing script artifacts and look for a most recent transaction on given chain_id which deployed contract matching the name